src/core/thread/mle.cpp

```diff
@@ -325,13 +325,22 @@ Error Mle::Restore(void)
 
     SuccessOrExit(error = Get<Settings>().Read(networkInfo));
 
+#if OPENTHREAD_MTD
+    VerifyOrExit(!IsActiveRouter(networkInfo.GetRloc16()), error = kErrorInvalidState);
+#endif
+
     Get<KeyManager>().SetCurrentKeySequence(networkInfo.GetKeySequence());
     Get<KeyManager>().SetMleFrameCounter(networkInfo.GetMleFrameCounter());
     Get<KeyManager>().SetAllMacFrameCounters(networkInfo.GetMacFrameCounter());
+
+#if OPENTHREAD_MTD
+    mDeviceMode.Set(networkInfo.GetDeviceMode() & ~DeviceMode::kModeFullThreadDevice);
+#else
     mDeviceMode.Set(networkInfo.GetDeviceMode());
+#endif
 
     // force re-attach when version mismatch.
-    VerifyOrExit(networkInfo.GetVersion() == kThreadVersion);
+    VerifyOrExit(networkInfo.GetVersion() == kThreadVersion, error = kErrorInvalidState);
```

Comment:
I think it was intentional to ignore setting error so that a device can successfully restore data after upgrading Thread version (e.g. by firmware update).

Reply:
I agree. Design-wise it sounds better to keep the error as `kErrorNone` in this case (since we still read and restore other stuff). The errors documented for `Restore()` in `mle.hpp` are currently `kErrorNone` or `kErrorNotFound`:
https://github.com/openthread/openthread/blob/3350f598f6759fd7de0c80845a925197458d0420/src/core/thread/mle.hpp#L146-L153


---

From what I see  The `Mle::Restore()` seems to be used only from `Instane::AfterInit()` and we do `IgnoreError()`. So maybe even better to change it to `void`.

Reply:
Agree to change to `void`.

Reply:
I think may be better to do the change to `void` from a separate PR (and not have it be tied to this PR which is addressing the  FTD to MTD firmware transition). I can submit a PR to change this soon.